STYLE.md

```diff
@@ -0,0 +1,262 @@
+
+While there are popular alternatives such as [`failure`](https://docs.rs/failure) and
+[`thiserror`](https://docs.rs/thiserror), they generally lack either the thoroughness in documentation or the
+flexibility of `snafu`.
+
+For **working with errors**, we have a more lax approach. At the highest level, we use `Box<dyn std::error::Error + Send
++ Sync + 'static>` for maximum flexibility. This allows developers to avoid needing to _always_ derive custom error
+  types in order to return errors back up the call stack. This does not prevent, and indeed, should not discourage
+  developers from using `snafu` to create rich error types that provide additional context, whether through descriptive
+  error messages, source errors, or backtraces.
+
+## Concurrency and synchronization
+
+### Atomics
+
+In general, we strive to use the atomic types in the [standard
+library](https://doc.rust-lang.org/stable/std/sync/atomic/index.html) when possible, as they are the most portable and
+well-tested. In cases where the standard library atomics cannot be used, such as when using a 64-bit atomic but wanting
+to support a 32-bit platform, or support a platform without atomic instructions at all, we prefer to use
+**[`crossbeam-utils`](https://docs.rs/crossbeam-utils)** and its `AtomicCell` helper. This type will automatically
+handle either using native atomic support or wrapping access in a mutex, and handle it in a transparent way. It uses a
+fixed acquire/release ordering that generally provides the expected behavior when using atomics, but may not be suitable
+for usages which require stronger ordering.
+
+### Global state
+
+When there is a need or desire to share global state, there are a few options depending on the required constraints.
+
+If you're working with data that is lazily initialized but never changes after initialization, we prefer
+**[`once_cell`](https://docs.rs/once_cell)**. It is slightly faster than [`lazy_static`], and additionally provides a
+richer API than both `lazy_static` and the standard library variants, such as `std::sync::Once`. Additionally, there is
+[active work happening](https://github.com/rust-lang/rust/issues/74465) to migrate the types in `once_cell` into
+`std::sync` directly, which will be easier to switch to if we're already using `once_cell`.
+
+If you're working with data that changes over time, but has a very high read-to-write ratio, such as many readers, but
+one writer and infrequent writes, we prefer **[`arc-swap`](https://docs.rs/arc-swap)**.  The main feature of this crate
+is allowing a piece of data to be atomically updated while being shared concurrently. It does this by wrapping all data
+in `Arc<T>` to provide the safe, concurrent access, while adding the ability to atomically swap the `Arc<T>` itself. As
+it cannot be constructed in a const fashion, `arc-swap` pairs well with `once_cell` for actually storing it in a global
+static variable.
+
+### Concurrent data structures
```

Comment:
If we're going to talk about these, would it also be worth mentioning `Mutex` vs `RwLock` and the std vs parking_lot vs tokio implementations?

Reply:
It probably would be a good idea, yes. I'll slot something in.